### PR TITLE
log error in link connectivity inside network routes

### PR DESCRIPTION
**Pull request by bhavyalatha26**
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/1759)

### CODE REVIEW

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java [hunk 1]
```diff
@@ -116,6 +116,23 @@ private void preparePhysSimForNewIteration() {
...
     }
 
+    private void validateForInvalidNodeChains() {
+        List<Link> errorLinks = new ArrayList<>();
+        Map<Id<Link>, ? extends Link> links = agentSimScenario.getNetwork().getLinks();
+        agentSimScenario.getPopulation().getPersons().values().
+                forEach(f -> f.getSelectedPlan().getPlanElements().forEach(pe -> {
+                    if(pe instanceof Leg) {
+                        Leg leg = (Leg) pe;
+                        Link start = links.get(leg.getRoute().getStartLinkId());
```
Comment (JustinPihony) [inline src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java]: From the description of #1731 it seems that this should be comparing two separate links, not the start and end of the same one. More like the end of the first, and the start of the next. Also, if there is any way to perform this check other than a double `for` that would be preferred for performance reasons.

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java [hunk 2]
```diff
@@ -116,6 +117,58 @@ private void preparePhysSimForNewIteration() {
...
+                            if(routeLinks.get(i).getToNode() != routeLinks.get(i+1).getFromNode()){
+                                brokenLinks.add(routeLinks.get(i).getId().toString() + "->" + routeLinks.get(i+1).getId().toString());
+                            }
+                        }
+                    }
+                }));
+        if(!brokenLinks.isEmpty()) {
+            StringBuilder brokenPaths = new StringBuilder();
+            brokenLinks.forEach(l -> brokenPaths.append(l).append(","));
+            log.info("Broken link paths observed in network route : " + brokenPaths.toString());
```
Comment (wrashid) [inline src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java]: just print number of broken routes on info level, detailed routes can be printed with debug level

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java [hunk 3]
```diff
@@ -116,6 +117,58 @@ private void preparePhysSimForNewIteration() {
...
+        Collection<? extends Link> links = agentSimScenario.getNetwork().getLinks().values();
+        for (Link link : links) {
+            if(link.getFromNode() == link.getToNode()){
+                errorLinks.add(link.getId().toString());
+            }
+        }
+        if(!errorLinks.isEmpty()) {
+            StringBuilder errorLinkIds = new StringBuilder();
+            errorLinks.forEach(id -> errorLinkIds.append(id).append(","));
+            log.info("Links that have same from and to nodes: " + errorLinkIds.toString());
```
Comment (wrashid) [inline src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java]: just print number of links with same from and to node, linkIds can be printed with debug level

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java [hunk 4]
```diff
@@ -142,6 +195,10 @@ private void setupActorsAndRunPhysSim(int iterationNumber) {
         linkStatsGraph.notifyIterationStarts(jdeqsimEvents, agentSimScenario.getConfig().travelTimeCalculator());
 
         log.info("JDEQSim Start");
+
+        checkForBrokenLinksInsideNetwork(jdeqSimScenario);
+        checkForErroneousLinkNodes();
```
Comment (wrashid) [inline src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java]: checkForErroneousLinkNodes - this check can be done just once - no need to do in every iteration.

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java [hunk 5]
```diff
@@ -116,6 +117,58 @@ private void preparePhysSimForNewIteration() {
         jdeqsimPopulation = PopulationUtils.createPopulation(agentSimScenario.getConfig());
     }
 
+    /**
+     * Logs all the broken links (if any) observed with a path inside a network route.
+     */
+    private void checkForBrokenLinksInsideNetwork(MutableScenario jdeqSimScenario) {
```
Comment (wrashid) [inline src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java]: rename checkForBrokenNetworkRoutes

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java [hunk 6]
```diff
@@ -116,6 +117,58 @@ private void preparePhysSimForNewIteration() {
         jdeqsimPopulation = PopulationUtils.createPopulation(agentSimScenario.getConfig());
     }
 
+    /**
+     * Logs all the broken links (if any) observed with a path inside a network route.
```
Comment (wrashid) [inline src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java]: reports routes, which are not following road network

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java [hunk 7]
```diff
@@ -116,6 +117,58 @@ private void preparePhysSimForNewIteration() {
...
+            StringBuilder brokenPaths = new StringBuilder();
+            brokenLinks.forEach(l -> brokenPaths.append(l).append(","));
+            log.info("Broken link paths observed in network route : " + brokenPaths.toString());
+        }
+    }
+
+    /**
+     * Logs all the erroneous links that have same from and to nodes.
+     */
+    private void checkForErroneousLinkNodes() {
```
Comment (wrashid) [inline src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java]: rename: checkForLinksWithSameFromAndToNode

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java (hunk 2 above)
Comment (bhavyalatha26) [reply src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java]: done

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java (hunk 3 above)
Comment (bhavyalatha26) [reply src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java]: done

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java (hunk 5 above)
Comment (bhavyalatha26) [reply src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java]: done

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java (hunk 6 above)
Comment (bhavyalatha26) [reply src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java]: done

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java (hunk 7 above)
Comment (bhavyalatha26) [reply src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java]: done

File: src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java (hunk 4 above)
Comment (bhavyalatha26) [reply src/main/java/beam/physsim/jdeqsim/AgentSimToPhysSimPlanConverter.java]: done

**Comment by JustinPihony** [comment]
Test!

**Pull request closed without merging**